common/src/main/java/io/pravega/common/io/filesystem/FileModificationPollingMonitor.java

```diff
@@ -0,0 +1,138 @@
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     *
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch}  or {@code callback} is null
+     */
+    public FileModificationPollingMonitor(@NonNull String fileToWatch, @NonNull Consumer<File> callback)
+            throws FileNotFoundException {
+        this(fileToWatch, callback, DEFAULT_POLL_INTERVAL);
+    }
+
+    @VisibleForTesting
+    FileModificationPollingMonitor(@NonNull String fileToWatch, @NonNull Consumer<File> callback, int pollingInterval)
+            throws FileNotFoundException {
+        this.pathOfFileToWatch = Paths.get(fileToWatch);
+        this.validateInput(this.pathOfFileToWatch, callback);
+        this.callback = callback;
+        this.pollingInterval = pollingInterval;
+    }
+
+    @Override
+    public void startMonitoring() {
+        Path fileName = pathOfFileToWatch.getFileName();
+        log.debug("fileName obtained from pathOfFileToWatch is [{}]", fileName);
+        if (fileName == null) {
+            throw new IllegalStateException("fileName is 'null'");
+        }
+
+        Path dirPath = pathOfFileToWatch.getParent();
+        log.debug("dirPath is [{}]", dirPath);
+        if (dirPath == null) {
+            throw new IllegalStateException("The directory containing the file turned out to be 'null`");
+        }
+
+        FileAlterationObserver observer = new FileAlterationObserver(dirPath.toString(), new FileFilter() {
+            @Override
+            public boolean accept(File file) {
+                if (file == null || file.getName() == null) {
+                    return false;
+                }
+                return file.getName().equals(fileName.toString());
+            }
+        });
+
+        monitor = new FileAlterationMonitor(this.pollingInterval);
+
+        FileAlterationListener listener = new FileAlterationListenerAdaptor() {
+            @Override
+            public void onFileChange(File file) {
+                log.info("Detected that the file [{}] has modified", file.getPath());
+                callback.accept(file);
+            }
+        };
+
+        observer.addListener(listener);
+        monitor.addObserver(observer);
+        try {
+            monitor.start();
+            log.info("Done setting up file modification monitor for file [{}]", this.pathOfFileToWatch);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void stopMonitoring() {
+        if (monitor != null) {
+            try {
+                monitor.stop(5 * 1000);
```

Comment:
This method will interrupt the current thread if the timeout expires. This will happen regardless of whether you catch it or not. As a result, whomever else is running on this thread will not be able to execute subsequent code.

Reply:
To test your assertion, I temporarily reduced the timeout to 1ms, and then ran a sample shown below that invoked stopMonitoring. I didn't observe the behavior you are mentioning: `exiting`. Am I missing something?

```
 @Test
    public void test()  {
        try {
            Path dir = Files.createTempDirectory("fw-");
            File file = File.createTempFile("tf-", ".temp", dir.toFile());

            AtomicBoolean isCallbackInvoked = new AtomicBoolean(false);
            FileModificationMonitor watcher = new FileModificationPollingMonitor(file.toPath().toString(),
                    c -> isCallbackInvoked.set(true), 100);
            watcher.startMonitoring();
            Thread.sleep(2 * 1000);
            watcher.stopMonitoring();
            System.out.println("exiting");
        } catch(Exception e) {
            e.printStackTrace();
        }
```